services-hll/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/model/ItemAttributeValue.java

```diff
@@ -0,0 +1,698 @@
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.core.util.SdkAutoConstructList;
+import software.amazon.awssdk.core.util.SdkAutoConstructMap;
+import software.amazon.awssdk.enhanced.dynamodb.converter.ItemAttributeValueConverter;
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * A simpler, and more user-friendly version of the generated {@link AttributeValue}.
+ *
+ * This is a union type of the types exposed by DynamoDB, exactly as they're exposed by DynamoDB.
+ *
+ * An instance of {@link ItemAttributeValue} represents exactly one DynamoDB type, like String (s), Number (n) or Bytes (b). This
+ * type can be determined with the {@link #type()} method or the {@code is*} methods like {@link #isString()} or
+ * {@link #isNumber()}. Once the type is known, the value can be extracted with {@code as*} methods like {@link #asString()}
+ * or {@link #asNumber()}.
+ *
+ * When converting an {@link ItemAttributeValue} into a concrete Java type, it can be tedious to use the {@link #type()} or
+ * {@code is*} methods. For this reason, a {@link #convert(TypeConvertingVisitor)} method is provided that exposes a polymorphic
+ * way of converting a value into another type.
+ *
+ * An instance of {@link ItemAttributeValue} is created with the {@code from*} methods, like {@link #fromString(String)} or
+ * {@link #fromNumber(String)}.
+ */
+@SdkPublicApi
+@ThreadSafe
+@Immutable
+public final class ItemAttributeValue {
+    private final ItemAttributeValueType type;
+    private final boolean isNull;
+    private final Map<String, ItemAttributeValue> mapValue;
+    private final String stringValue;
+    private final String numberValue;
+    private final SdkBytes bytesValue;
+    private final Boolean booleanValue;
+    private final List<String> setOfStringsValue;
+    private final List<String> setOfNumbersValue;
+    private final List<SdkBytes> setOfBytesValue;
+    private final List<ItemAttributeValue> listOfAttributeValuesValue;
+
+    private ItemAttributeValue(InternalBuilder builder) {
+        this.type = builder.type;
+        this.isNull = builder.isNull;
+        this.stringValue = builder.stringValue;
+        this.numberValue = builder.numberValue;
+        this.bytesValue = builder.bytesValue;
+        this.booleanValue = builder.booleanValue;
+
+        this.mapValue = builder.mapValue == null
+                        ? null
+                        : Collections.unmodifiableMap(new LinkedHashMap<>(builder.mapValue));
```

Comment:
Can we use `CollectionUtils.deepCopyMap`?

Reply:
Unless I am mistaken, deepCopyMap is for Map<X, List<Y>>, not Map<X, Y>.